File: algorithm/acwing/LanQiaoBeiDayOne2023/src/Week1/day5/AcWing3777/main.go
package main

import (
    "bufio"
    "fmt"
    "os"
)
/**
 * @Author 袁佳林
 * @Description
 * @Date 2023/2/24 21:03
 */
//N const
var n, m, t int
var s string
var chs []byte
var out = bufio.NewWriter(os.Stdout)
var in = bufio.NewReader(os.Stdin)
func solve() {
    fmt.Fscan(in, &n, &s)
    chs = []byte(s)
    for i := 0; i < n; i++ {
        if chs[i] == 'W' {
            chs[i] = 1
        } else {
            chs[i] = 0
        }
    }
    one, zero := 0, 0
    for i := 0; i < n; i++ {
        if chs[i] == 0 {
            zero++
        } else {
            one++
        }
    }
    if zero%2 == 1 && one%2 == 1 {
        fmt.Fprintln(out, -1)
        return
    }
    if one%2 == 0 {
        work(0)
    } else {
        work(1)
    }
}

func work(c byte) {
    res := []int{}
    for i := 0; i < n - 1; i++ {
        if chs[i] != c {
            res = append(res, i + 1)
            chs[i] ^= 1
            chs[i + 1] ^= 1
        }
    }
    fmt.Fprintln(out, len(res))
    for _, v := range res {
        fmt.Fprintf(out, "%d ", v)
    }
    if len(res) != 0 {
        fmt.Fprintln(out)
    }
}

func main() {
    defer out.Flush()
    for fmt.Fscan(in, &t); t > 0; t-- {
        solve()
    }
}

File: algorithm/acwing/LanQiaoBeiDayOne2023/src/Week1/day5/AcWing3777/main.cpp
#include <iostream>
#include <cstring>
#include <algorithm>
#include <vector>
/**
 * @Author 袁佳林
 * @Description
 * @Date 2023/2/24 21:03
 */
using namespace std;

int n;
string str;

void update(char& c)
{
    if (c == 'W') c = 'B';
    else c = 'W';
}

bool check(char c)
{
    vector<int> res;
    string s = str;
    for (int i = 0; i + 1 < n; i ++ )
        if (s[i] != c)
        {
            update(s[i]);
            update(s[i + 1]);
            res.push_back(i);
        }

    if (s.back() != c) return false;

    cout << res.size() << endl;
    for (int x: res) cout << x + 1 << ' ';
    if (res.size()) cout << endl;

    return true;
}

int main()
{
    int T;
    cin >> T;
    while (T -- )
    {
        cin >> n >> str;
        if (!check('B') && !check('W')) puts("-1");
    }

    return 0;
}

